gno.land/pkg/gnoland/app.go

```diff
@@ -108,14 +99,28 @@ func NewAppWithOptions(cfg *AppOptions) (abci.Application, error) {
 			newCtx sdk.Context, res sdk.Result, abort bool,
 		) {
 			// Override auth params.
-			ctx = ctx.WithValue(
-				auth.AuthParamsContextKey{}, auth.DefaultParams())
+			ctx = ctx.
+				WithValue(auth.AuthParamsContextKey{}, auth.DefaultParams())
 			// Continue on with default auth ante handler.
 			newCtx, res, abort = authAnteHandler(ctx, tx, simulate)
 			return
 		},
 	)
 
+	// Set begin and end transaction hooks.
+	// These are used to create gno transaction stores and commit them when finishing
+	// the tx - in other words, data from a failing transaction won't be persisted
+	// to the gno store caches.
+	baseApp.SetBeginTxHook(func(ctx sdk.Context) sdk.Context {
+		// Create Gno transaction store.
+		return vmk.MakeGnoTransactionStore(ctx)
+	})
+	baseApp.SetEndTxHook(func(ctx sdk.Context, result sdk.Result) {
+		if result.IsOK() {
+			vmk.CommitGnoTransactionStore(ctx)
+		}
+	})
```

Comment:
I still have to go through more code, but should we rollback if the result is not ok to avoid more cache/inconsistencies problems?

Reply:
If that is the case, how we are discarding previously stored data in memory, creating a new transaction, and/or expecting the GC to clean it?

Reply:
huh?

The variables fall out of scope and the memory is retrieved by the GC. This happens in `BaseApp.runTx`; `cacheTxContext` does the "cache" wrapping (= tx log) around the stores; if `msCache.MultiWrite()` is never called, then `msCache` falls out of scope and is then collected by the GC.

Similarly is done for the Gno TransactionStore